examples/parallel_wavegan/train_parallel_wavegan.py

```diff
@@ -0,0 +1,470 @@
+        dict_metrics_losses.update({"spectral_convergence_loss": sc_loss})
+        dict_metrics_losses.update({"log_magnitude_loss": mag_loss})
+
+        per_example_losses = gen_loss
+        return per_example_losses, dict_metrics_losses
+
+    def compute_per_example_discriminator_losses(self, batch, gen_outputs):
+        audios = batch["audios"]
+        y_hat = gen_outputs
+
+        y = tf.expand_dims(audios, 2)
+        p = self._discriminator(y)
+        p_hat = self._discriminator(y_hat)
+
+        real_loss = 0.0
+        fake_loss = 0.0
+
+        real_loss += calculate_3d_loss(tf.ones_like(p), p, loss_fn=self.mse_loss)
+        fake_loss += calculate_3d_loss(
+            tf.zeros_like(p_hat), p_hat, loss_fn=self.mse_loss
+        )
+
+        dis_loss = real_loss + fake_loss
+
+        # calculate per_example_losses and dict_metrics_losses
+        per_example_losses = dis_loss
+
+        dict_metrics_losses = {
+            "real_loss": real_loss,
+            "fake_loss": fake_loss,
+            "dis_loss": dis_loss,
+        }
+
+        return per_example_losses, dict_metrics_losses
+
+    def generate_and_save_intermediate_result(self, batch):
+        """Generate and save intermediate result."""
+        import matplotlib.pyplot as plt
+
+        # generate
+        y_batch_ = self.one_step_predict(batch)
+        y_batch = batch["audios"]
+
+        # convert to tensor.
+        # here we just take a sample at first replica.
+        try:
+            y_batch_ = y_batch_.values[0].numpy()
+            y_batch = y_batch.values[0].numpy()
+        except Exception:
+            y_batch_ = y_batch_.numpy()
+            y_batch = y_batch.numpy()
+
+        # check directory
+        dirname = os.path.join(self.config["outdir"], f"predictions/{self.steps}steps")
+        if not os.path.exists(dirname):
+            os.makedirs(dirname)
+
+        for idx, (y, y_) in enumerate(zip(y_batch, y_batch_), 1):
+            # convert to ndarray
+            y, y_ = tf.reshape(y, [-1]).numpy(), tf.reshape(y_, [-1]).numpy()
+
+            # plit figure and save it
+            figname = os.path.join(dirname, f"{idx}.png")
+            plt.subplot(2, 1, 1)
+            plt.plot(y)
+            plt.title("groundtruth speech")
+            plt.subplot(2, 1, 2)
+            plt.plot(y_)
+            plt.title(f"generated speech @ {self.steps} steps")
+            plt.tight_layout()
+            plt.savefig(figname)
+            plt.close()
+
+            # save as wavefile
+            y = np.clip(y, -1, 1)
+            y_ = np.clip(y_, -1, 1)
+            sf.write(
+                figname.replace(".png", "_ref.wav"),
+                y,
+                self.config["sampling_rate"],
+                "PCM_16",
+            )
+            sf.write(
+                figname.replace(".png", "_gen.wav"),
+                y_,
+                self.config["sampling_rate"],
+                "PCM_16",
+            )
+
+
+def main():
```

Comment:
Move to click from argparse https://click.palletsprojects.com/en/7.x/why/#
A lot cleaner and easier to modify.